book/src/functional-testing.md

```diff
@@ -1,7 +1,84 @@
+Drivers should be functionally "end-to-end" tested while deployed in a Kubernetes cluster. Previously how to do this and what tests to run was left up to driver authors. Now, a standard set of Kubernetes CSI end-to-end tests can be imported and run by third party CSI drivers. This documentation specifies how to do so.
 
 The CSI community is also looking in to establishing an official "CSI Conformance Suite" to recognize "officially certified CSI drivers".  This documentation will be updated with more information once that process has been defined.
+
+# Kubernetes End to End Testing for CSI Storage Plugins
+
+Currently, [csi-sanity](https://github.com/kubernetes-csi/csi-test/tree/master/cmd/csi-sanity) exists to help test compliance with the CSI spec, but e2e testing of plugins is needed as well to provide plugin authors and users validation that their plugin is integrated well with specific versions of Kubernetes.
+
+## Setting up End to End tests for your CSI Plugin
+
+### Prerequisites:
+ * A Kubernetes v1.13+ Cluster
+ * [Kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/#install-kubectl)
+
+There are two ways to run end-to-end tests for your CSI Plugin
+ 1) use [Kubernetes E2E Tests](https://github.com/kubernetes/kubernetes/tree/release-1.14/test/e2e/storage/external), by providing a DriverDefinition YAML file via a parameter. 
+ * **Note**: In some cases you would not be able to use this method, in running e2e tests by just providing a YAML file defining your CSI plugin. For example the [NFS CSI plugin](https://github.com/kubernetes-csi/csi-driver-nfs) does not support dynamic provisoning, so we would want to skip those and run only pre-provisioned tests. For such cases, you would need to write your own testdriver, which is discussed below. 
+ 
+ 2) import the in-tree storage tests and run them using `go test`. 
+ 
+ This doc will cover how to run the E2E tests using the second method.
+
+## Importing the E2E test suite as a library
+
+In-tree storage e2e tests could be used to test CSI storage plugins. Your repo should be setup similar to how the [NFS CSI plugin](https://github.com/kubernetes-csi/csi-driver-nfs/) is setup, where the testfiles are in a `test` directory and the main test file is in the `cmd` directory. 
+
+To be able import Kubernetes in-tree storage tests, the CSI plugin would need to use **Kubernetes v1.14+** (add to plugin's GoPkg.toml, since pluggable E2E tests become available in v1.14). CSI plugin authors would also be required to implement a [testdriver](https://github.com/kubernetes/kubernetes/blob/release-1.14/test/e2e/storage/testsuites/testdriver.go#L31) for their CSI plugin. The testdriver provides required functionality that would help setup testcases for a particular plugin. 
```

Comment:
Can you actually use a specific commit hash or tag in the url so that it's not possible for the line number to change?

Same for all links to line numbers

Reply:
@msau42 I've updated the links, can you check if they're okay now?